diff --git a/usr.sbin/bsdinstall/bsdinstall b/usr.sbin/bsdinstall/bsdinstall
index 19e27c929aee..80972c655ba9 100755
--- a/usr.sbin/bsdinstall/bsdinstall
+++ b/usr.sbin/bsdinstall/bsdinstall
@@ -76,7 +76,7 @@ if [ "$debug" ]; then
 	"") exec "/usr/libexec/bsdinstall/$VERB" "$@" 2> /dev/null ;;
 	# If begins with `+', send errors to both terminal and file (no `+')
 	+*) exec "/usr/libexec/bsdinstall/$VERB" "$@" \
-		2>&1 >&$TERMINAL_STDOUT_PASSTHRU | tee "${debugFile#+}" ;;
+		2>&1 >&$TERMINAL_STDOUT_PASSTHRU | tee -a "${debugFile#+}" ;;
 	# Otherwise, just send errors to the file specified
 	*) exec "/usr/libexec/bsdinstall/$VERB" "$@" 2>> "$debugFile"
 	esac
diff --git a/usr.sbin/bsdinstall/scripts/script b/usr.sbin/bsdinstall/scripts/script
index ae1c6b3011fa..7f5ec1c38f56 100755
--- a/usr.sbin/bsdinstall/scripts/script
+++ b/usr.sbin/bsdinstall/scripts/script
@@ -141,7 +141,7 @@ else
 		# case of EFI system partitions only. This *ONLY WORKS* if
 		# /boot/efi is empty and does not handle analagous problems on
 		# other systems (ARM, PPC64).
-		tar -xf "$BSDINSTALL_DISTDIR/$set" -C $BSDINSTALL_CHROOT --exclude boot/efi
+		tar -xPf "$BSDINSTALL_DISTDIR/$set" -C $BSDINSTALL_CHROOT --exclude boot/efi
 		mkdir -p $BSDINSTALL_CHROOT/boot/efi
 	done
 fi
diff --git a/usr.sbin/bsdinstall/scripts/zfsboot b/usr.sbin/bsdinstall/scripts/zfsboot
index a671841b15e5..2e72a3165409 100755
--- a/usr.sbin/bsdinstall/scripts/zfsboot
+++ b/usr.sbin/bsdinstall/scripts/zfsboot
@@ -1463,6 +1463,9 @@ zfs_create_boot()
 	             "$zroot_name" || return $FAILURE
 
 	if [ "$ZFSBOOT_BOOT_POOL" ]; then
+		f_eval_catch $funcname zpool "$ZPOOL_SET" \
+			"cachefile=\"$BSDINSTALL_CHROOT/boot/zfs/zpool.cache\"" \
+			"$bootpool_name" || return $FAILURE
 		f_eval_catch $funcname printf "$PRINTF_CONF" \
 			vfs.root.mountfrom \
 			"\"zfs:$zroot_name/$zroot_bootfs\"" \
@@ -1511,6 +1514,26 @@ zfs_create_boot()
 		             return $FAILURE
 	fi
 
+	if [ "$ZFSBOOT_BOOT_POOL" ]; then
+		local cachepath="$bootpool/boot/zfs/zpool.cache"
+		local cachedir="$BSDINSTALL_CHROOT/var/cache"
+		local cachefile="$cachedir/zpool.cache"
+
+		f_eval_catch $funcname mkdir "$MKDIR_P" $cachedir ||
+			     return $FAILURE
+		f_eval_catch $funcname cp "$COPY" $cachepath $cachedir ||
+			     return $FAILURE
+		f_eval_catch $funcname zpool "$ZPOOL_EXPORT" $bootpool_name ||
+			     return $FAILURE
+		f_eval_catch $funcname mkdir "$MKDIR_P" $bootpool/boot/zfs ||
+			     return $FAILURE
+		f_eval_catch $funcname cp "$COPY" $cachefile $cachepath ||
+			     return $FAILURE
+		f_eval_catch $funcname zpool "$ZPOOL_IMPORT_WITH_OPTIONS" \
+			     "-o altroot=\"$BSDINSTALL_CHROOT\"" \
+			     $bootpool_name || return $FAILURE
+	fi
+
 	# We're all done unless we should go on to do encryption
 	[ "$ZFSBOOT_GELI_ENCRYPTION" ] || return $SUCCESS
 
